File: smallblog.pl
#!/usr/bin/perl
use strict;
use warnings;
use Template;
use YAML::Tiny;
use File::Slurp;
use Text::Markdown::Discount qw(markdown);
use File::stat;
use POSIX qw(strftime);

my $yaml = YAML::Tiny->read('smallblog.conf');
my $site = $yaml->[0];

my $tt = Template->new({
	INCLUDE_PATH => 'templates',
	INTERPOLATE => 1,					# Shell style ${...} variables
	#PRE_CHOMP => 1,
	POST_CHOMP => 1,
}) || die "$Template::ERROR\n";

# Takes a path to a post.md and returns a $vars hash
sub parse_post {
	my $path = shift;
	my $text = read_file($path);
	my $title;
	my @tags;
	my $date;
	my $edit_date;
	my $html;

	# Grab title
	$text =~ s/^# (?<title>.+)//;
	$title = $+{title};

	# Grab tags into an array
	$text =~ s/^tags: (?<tags>.+?)\n//ms;
	# Check if we have tags
	if ($+{tags}) {
		@tags = split(" ", $+{tags});
		# Update the global tags list
		$site->{tags}{$_} = 1 foreach (@tags);
	}

	# Grab date from path and reformat slashes to dashes
	my ($year,$month,$day) = ($path =~ m/(.*)\/(.*)\/(.*)\/.*.md/);
	$date = "$year-$month-$day";

	# Grab the file mtime with stat as edited time
	my $st = stat($path);
	my $fdt = strftime("%Y-%m-%d", localtime($st->mtime));
	$edit_date = $fdt if ($date ne $fdt);

	$html = markdown($text);

	# Local (post-specific) variables
	return {
		path		=> $path,
		title		=> $title,
		tags		=> \@tags,
		date		=> $date,
		edit_date	=> $edit_date,
		html		=> $html,
	};
}

# Grab markdown files from a YYYY/MM/DD/post.md structure
my @paths = split("\n", qx(ls -r */*/*/*.md));

# Prase the files into an array to save filesystem lookups
my @posts = map {parse_post($_)} @paths;

# Generate posts.md.html
foreach my $post (@posts) {
	my $vars = {
		site => $site,
		post => $post,
	};
	$tt->process('post_page.tmpl', $vars, "$post->{path}.html")
		|| die "Generating $post->{path} failed: ", $tt->error(), "\n";
}

# Generate index.html
{
	#my $limit = ${@posts} > $site->{max_posts} ? ${@posts} : $site->{max_posts};
	if ($site->{max_posts} > scalar @posts) {
		$site->{max_posts} = scalar @posts;
	}

	my $vars = {
		site => $site,
		posts => [@posts[0 .. $site->{max_posts}-1]],
	};
	$tt->process('index.tmpl', $vars, "index.html")
		|| die "index.html failed: ", $tt->error(), "\n";
}

# Generate archive.html
{
	my $vars = {
		site => $site,
		posts => \@posts,
		title => 'Archives',
	};
	$tt->process('archive_page.tmpl', $vars, "archive.html")
		|| die "archive.html failed: ", $tt->error(), "\n";
}

# Generate tags/index.html
{
	my $vars = {
		site => $site,
		title => 'Tags',
	};
	$tt->process('tag_index.tmpl', $vars, "tags.html")
		|| die "tags/index.html failed: ", $tt->error(), "\n";
}

# Generate tags/$tag.html
foreach my $tag (keys $site->{tags}) {
	my @tag_posts = map {grep(/$tag/, @{$_->{tags}}) ? $_ : ()} @posts;

	my $vars = {
		site => $site,
		posts => \@tag_posts,
		title => $tag,
	};
	$tt->process('tag_page.tmpl', $vars, "tags/$tag.html")
		|| die "tags/$tag.html failed: ", $tt->error(), "\n";
} if ($site->{tags})

# Generate feed.rss
{
	my $vars = {
		site => $site,
		posts => \@posts,
	};
	$tt->process('rss.tmpl', $vars, "feed.rss")
		|| die "feed.rss failed: ", $tt->error(), "\n";
}

File: smallblog.lua
-- TODO: 
--  - Error out if tags/ doesn't exist
--
local posix		= require "posix"				-- LuaPosix
local sys_stat	= require "posix.sys.stat"	-- LuaPosx
local yaml		= require "lyaml"
local markdown	= require "discount"
local template	= require "resty.template"

local utils = {}
-- open and read a file into a string
function utils.readFile( path )
	local f			= assert( io.open( path, 'r' ) )
	local contents	= assert( f:read( '*a' ) )
	f:close()
	return contents
end
-- open and write a file from a string
function utils.writeFile( path, s )
	local f = assert( io.open( path, 'w+' ) )
	assert( f:write( s ) )
	f:close()
	return true
end
-- string.gsub with captures
-- returns capture,original
function utils.cgsub( s, pattern, replace )
	c = s:match( pattern )
	s = s:gsub( pattern, replace )
	return c,s
end
-- sort a hash table into an array
function utils.tsort( t )
	local array = {}
	for k,v in pairs( t ) do table.insert( array, k ) end
	table.sort( array )
	return array
end

local function parsePost( path )
	-- Post data structure
	local tags = ""
	local post = {
		path = path,
		text	= "",		html = "",
		title	= "",		tags = {},
		date	= "", 	edit_date = "",
	}

	-- Grab date from path
	local year, month, day = path:match( "(%d+)/(%d+)/(%d+)/.*%.md" )
	-- Ignore files not in the proper hierarchy
	if year == nil or month == nil or day == nil then return nil end
	-- Grab the mtime
	local fattr = sys_stat.stat( path )
	-- Format the dates to YYYY-MM-DD
	post.date		= string.format( "%s-%s-%s", year, month, day )
	post.edit_date	= os.date( "%Y-%m-%d", fattr.st_mtime )

	-- Read in post
	post.text = utils.readFile( path )

	-- Get post title
	post.title, post.text = utils.cgsub( post.text, "^# (.-)\n", "" )
	-- Get post tags
	tags, post.text = utils.cgsub( post.text, "tags: (.-)\n", "" )
	-- Index tags and append them to the global hash
	for tag in (tags or ""):gmatch( "%g+" ) do
		table.insert( post.tags, tag )
		site.tags[tag] = true
	end

	-- Convert the text to markdown
	post.html = markdown( post.text )

	-- Return the structure
	return post
end


-- Read in the site configuration
site = yaml.load( utils.readFile( "smallblog.conf" ) )
site.tags = {}
-- Grab all posts in the form of YYYY/MM/DD/post.md
local paths = posix.glob( "*/*/*/*.md" )
-- Sort posts in reverse order (newest first)
table.sort( paths, function(a,b) return a > b end )

-- Generate the individual posts.md.html and save posts
local posts			= {}
local indexPosts	= {}
local view			= template.new( "templates/post_page.tmpl" )
for _,path in ipairs( paths ) do
	post = parsePost( path )
	if #indexPosts < site.max_posts then table.insert( indexPosts, post ) end
	if post then
		table.insert( posts, post )

		view.site	= site
		view.post	= post
		utils.writeFile( path..".html", tostring( view ) )
	end
end

-- Generate index.html
do
	view			= template.new( "templates/index.tmpl" )
	view.site	= site
	view.posts	= indexPosts
	utils.writeFile( "index.html", tostring( view ) )
end

-- Generate archive.html
do
	view			= template.new( "templates/archive_page.tmpl" )
	view.site	= site
	view.posts	= posts
	view.title	= "Archives"
	utils.writeFile( "archive.html", tostring( view ) )
end

-- Generate tags.html
do
	view			= template.new( "templates/tag_index.tmpl" )
	-- Sort tags
	site.tags	= utils.tsort( site.tags )
	view.site	= site
	view.title	= "Tags"
	utils.writeFile( "tags.html", tostring( view ) )
end

-- Generate tags/$tag.html
view = template.new( "templates/tag_page.tmpl" )
for _,tag in ipairs( site.tags ) do
	-- TODO: clean out stale tags in tags/
	local tagList = {}
	for _,post in ipairs( posts ) do
		if table.concat( post.tags ):match( tag ) then
			table.insert( tagList, post )
		end
	end

	view.site	= site
	view.posts	= tagList
	view.title	= tag
	utils.writeFile( "tags/"..tag..".html", tostring( view ) )
end

-- Generate feed.rss
do
	view			= template.new( "templates/rss.tmpl" )
	view.site	= site
	view.posts	= posts
	utils.writeFile( "feed.rss", tostring( view ) )
end
